test/cpp/api/init.cpp

```diff
@@ -0,0 +1,88 @@
+#include <gtest/gtest.h>
+
+#include <torch/nn/init.h>
+#include <torch/nn/modules/linear.h>
+
+#include <test/cpp/api/init_baseline.h>
+#include <test/cpp/api/support.h>
+
+#include <vector>
+
+void check_exact_values(
+    std::vector<torch::Tensor> parameters,
+    std::vector<std::vector<torch::Tensor>> expected_parameters) {
+  ASSERT_EQ(parameters.size(), expected_parameters.size());
+
+  for (size_t i = 0; i < parameters.size(); i++) {
+    auto layerParameters = parameters[i];
+    auto expectedLayerParameters = expected_parameters[i];
+
+    ASSERT_EQ(layerParameters.size(0), expectedLayerParameters.size());
+
+    for (size_t p = 0; p < layerParameters.size(0); p++) {
+      auto tensor = layerParameters[p];
+      auto expectedTensor = expectedLayerParameters[p];
+
+      ASSERT_TRUE(
+          tensor.allclose(expectedTensor, /*rtol=*/1e-3, /*atol=*/5e-4));
+    }
+  }
+}
+
+void check_initializer_against_baseline(
+    std::function<void(torch::Tensor)> initializer,
+    std::vector<std::vector<torch::Tensor>> expected) {
+  torch::manual_seed(0);
+
+  auto layer1 = torch::nn::Linear(7, 15);
+  initializer(layer1->weight);
+  layer1->to(torch::kFloat64);
+
+  auto layer2 = torch::nn::Linear(15, 15);
+  initializer(layer2->weight);
+  layer2->to(torch::kFloat64);
+
+  auto layer3 = torch::nn::Linear(15, 2);
+  initializer(layer3->weight);
+  layer3->to(torch::kFloat64);
+
+  auto parameters = std::vector<torch::Tensor>{
+      layer1->weight,
+      layer2->weight,
+      layer3->weight,
+  };
+
+  check_exact_values(parameters, expected);
+}
+
+TEST(InitTest, ProducesPyTorchValues_XavierUniform) {
+  auto expected = expected_parameters::Xavier_Uniform();
+  auto initializer = [](torch::Tensor tensor) -> void {
+    torch::nn::init::xavier_uniform_(tensor);
```

Comment:
Actually why aren't you just passing `&torch::nn::init::xavier_uniform_`; your lambda here doesn't do anything.
`check_initializer_against_baseline(expected_parameters::Xavier_Uniform(), &torch::nn::init::xavier_uniform_);` should work

Reply:
I've never actually used C++ lambdas before, but I kept running into type errors when I tried that.

Xavier's signature is: 
```
TORCH_API Tensor xavier_normal_(
    Tensor tensor, 
    double gain = 1.0);
```
Kaiming's signature is:

```
TORCH_API Tensor kaiming_normal_(
    Tensor tensor,
    double a = 0,
    FanMode mode = FanMode::FanIn,
    Nonlinearity nonlinearity = Nonlinearity::LeakyReLU);
```

My test code's signature looks like:
```
void check_initializer_against_baseline(
    std::function<void(torch::Tensor)> initializer,
    std::vector<std::vector<torch::Tensor>> expected) 
```

When I try passing `kaiming_normal_` to my test code I get the following:
```
/home/josh/git/pytorch/test/cpp/api/init.cpp:90:80: error: could not convert ‘torch::nn::init::kaiming_normal_’ from ‘at::Tensor (*)(at::Tensor, double, torch::nn::init::FanMode, torch::nn::init::Nonlinearity)’ to ‘std::function<void(at::Tensor)>
```
Despite the `double`, `FanMode`, `Nonlinearity` etc. having default values, it seems like they're still taken into account  when passing them as an argument to a function. This is why I used lambdas with the same signature `Tensor -> void`.

Do you know a workaround for this? Maybe I've defined my `std::function<void(torch::Tensor)> initializer` incorrectly?

Reply:
Ah I think it won't work because of the default values. That's fine then.